llvm/lib/Target/AMDGPU/Utils/AMDGPUBaseInfo.cpp

```diff
@@ -986,7 +986,9 @@ unsigned getMaxFlatWorkGroupSize(const MCSubtargetInfo *STI) {
 
 unsigned getWavesPerWorkGroup(const MCSubtargetInfo *STI,
                               unsigned FlatWorkGroupSize) {
-  return divideCeil(FlatWorkGroupSize, getWavefrontSize(STI));
+  // divideCeil will overflow, unless FlatWorkGroupSize is cast.
+  return divideCeil(static_cast<uint64_t>(FlatWorkGroupSize),
+                    getWavefrontSize(STI));
```

Comment:
There's no way this should require 64-bit

Reply:
I was also surprised, but the assertion is hit in CodeGen/AMDGPU/codegen-internal-only-func.ll.

Orthogonal, but we could probably get divideCeil to never overflow, using some clever addition/subtraction.

Reply:
Hm, I think it's pretty unfortunate that we need these upcasts to avoid overflow. This would be a new footgun introduced by this change.

I think we should change the implementations to avoid overflow first (outside of this PR). Assuming this doesn't have measurable compile-time overhead (which I would not expect), it's best to make these function safe by default.

I think a typical way to do that is basically what the signed version does with `(X - 1) / Y + 1` -- it's just that this require an extra check for `X == 0` upfront. I guess a branchless variant of that would be `Bias = X != 0; return (X - Bias) / Y + Bias`.

Reply:
On a related note, how do I rewrite `alignTo`, which is `(Value + Align - 1) / Align * Align` to not overflow?

Reply:
Oh, right. There's no way to avoid overflowing, as the final `* Align` may overflow.